### ISSUE TITLE
SDI printf

### ISSUE THREAD

**Issue by ldab** [Community-PIO-CH32V/platform-ch32v#62]
I'm not too familiar with how things are set up here, but I have a `CH32V003` printing over SDI (one wire) but for that to happen first the "enable SDI Printf" needs to be sent on the WCH-LinkUtility, I can do so on Windows, but on linux the tool is not available:

![image](https://github.com/Community-PIO-CH32V/platform-ch32v/assets/34199302/fb97d1a1-70ff-4d88-a26c-12ec05cebf75)

It seems tho that a CLI is available here: https://github.com/ch32-rs/wlink#feature-support which seems to support the feature, so theoretically we sent `wlink flash --enable-sdi-print` do you see how this could be integrated on the flasher?

Basically the issues is more like a dumb question to "where/which flasher is used?"

**Comment by ldab** [comment]
Ok, now I see, https://github.com/Community-PIO-CH32V/wchisp is a fork of the repo mentioned above, do you think we can sync it?

**Comment by ldab** [comment]
So, theoretically I add a new uploader here? https://github.com/Community-PIO-CH32V/platform-ch32v/blob/6dce5eb1338cc96587d70c0bc814b8357378bd5a/platform.json#L142

**Comment by maxgerhardt** [comment]
>theoretically I add a new uploader here?

Yes. But that also has to be matched by providing the download URL for PlatformIO compatible packages for that tool (tar.gz packages that have `package.json` + the [platform-specific wlink executable](https://github.com/ch32-rs/wlink/releases/tag/nightly)). `wlink` is not yet added, but can easily be added.

To answer the question: 
>"where/which flasher is used?"

Depending on which `upload_protocol = ...` is selected from the available ones

https://github.com/Community-PIO-CH32V/platform-ch32v/blob/6dce5eb1338cc96587d70c0bc814b8357378bd5a/boards/ch32v307_evt.json#L44-L48

we either use
1. The WCH fork of *OpenOCD*.
2. CNLohr's `minichlink` executable ([here](https://github.com/Community-PIO-CH32V/ch32v003fun/tree/master/minichlink))
3. Or wchisp (only thing that can do USB ISP bootloader uploads now)

The [`builder.py`](https://github.com/Community-PIO-CH32V/platform-ch32v/blob/6dce5eb1338cc96587d70c0bc814b8357378bd5a/builder/main.py#L133-L179) code decides how to invoke these uploads then. The `platform.py` usually holds the download links for the custom packages.

Let me just add `wlink` tool support and a platform target for "View SDI Printfs via wlink".,

**Comment by maxgerhardt** [comment]
Ahh, I have to correct myself. We don't even need to enable uploading with that tool, just one-time invoking `wlink` with that `--enable-sdi-print` is going to make the serial port of the WCH-Link spit out the decoded SDI prints automatically, as normal?

**Comment by ldab** [comment]
> Ahh, I have to correct myself. We don't even need to enable uploading with that tool, just one-time invoking it `wlink` with that `--enable-sdi-print` is going to make the serial port of the WCH-Link spit out the decoded SDI prints automatically, as normal?

Agreed, after enabling, WCH-Link will spit the printfs.

I will add an example after that.

**Comment by maxgerhardt** [comment]
I've made addition https://github.com/Community-PIO-CH32V/platform-ch32v/commit/a73ed6d664f4cefff840a7f79228122fd82c874f.
* `upload_protocol = wlink` now available for all boards, will use `wlink flash` to flash
* `tool-wlink` now always installed from https://github.com/Community-PIO-CH32V/tool-wlink no matter the uploader
  * so that "Enable SDI Printf" task is available even if `upload_protocol = wch-link` (OpenOCD) is selected, otherwise the tool-wlink package may not be guaranteed to be installed
* task for "Enable SDI Printf" available, triggers `wlink sdi-print enable`

![grafik](https://github.com/Community-PIO-CH32V/platform-ch32v/assets/26485477/00dd39ea-1880-4b3b-90c7-a9aea2531fc9)

You'll need to do a full platform update and window reload of VSCode. Easiest is to delete `C:\Users\<user>\.platformio\platforms\ch32v*` and restart VSCode.

**Comment by ldab** [comment]
that is super cool, it works no problem:

```
Enabling SDI Printf
14:01:18 [INFO] Connected to WCH-Link v2.10(v30) (WCH-LinkE-CH32V305)
14:01:19 [INFO] Attached chip: CH32V003 [CH32V003F4P6] (ChipID: 0x00300500)
14:01:19 [INFO] Enabling SDI print
14:01:19 [INFO] Now you can connect to the WCH-Link serial port
```

![image](https://github.com/Community-PIO-CH32V/platform-ch32v/assets/34199302/9e250cb4-20b8-4454-b58e-0b023ea0da5f)

I think the next step would be to add an option to the monitor itself, something like `monitor_sdi = true` to the ini file, what do you think?

**Comment by maxgerhardt** [comment]
Would be nice, but I'm not sure at the moment if the PlatformIO core allows a "pre" hook for the `monitor` task or a prediction thereof, where we could execute a quick `wlink sdi-print enable` if that project option is set..
